packages/components/src/range-control/index.js

```diff
@@ -26,19 +27,48 @@ function RangeControl( {
 	help,
 	allowReset,
 	initialPosition,
+	min,
+	max,
+	setState,
 	...props
 } ) {
 	const id = `inspector-range-control-${ instanceId }`;
-	const resetValue = () => onChange();
+	const currentInputValue = currentInput === null ? value : currentInput;
+	const resetValue = () => {
+		resetCurrentInput();
+		onChange();
+	};
+	const resetCurrentInput = () => {
+		if ( currentInput !== null ) {
+			setState( {
+				currentInput: null,
+			} );
+		}
+	};
+
 	const onChangeValue = ( event ) => {
 		const newValue = event.target.value;
-		if ( newValue === '' ) {
-			resetValue();
+		const newNumericValue = parseInt( newValue, 10 );
+		// If the input value is invalid temporarly save it to the state,
+		// without calling on change.
+		if (
+			isNaN( newNumericValue ) ||
+			( min && newNumericValue < min ) ||
+			( max && newNumericValue > max )
```

Comment:
Depending if we wanted to support negative numbers, the following hypothetical unit test fails:

```js
it( 'validates when provided a max or min of zero', () => {
	const onChange = jest.fn();
	const wrapper = getWrapper( { onChange, min: -100, value: 0 } );

	const numberInputElement = () => TestUtils.findRenderedDOMComponentWithClass(
		wrapper,
		'components-range-control__number'
	);

	TestUtils.Simulate.change(
		numberInputElement(),
		{
			target: { value: '1' },
		}
	);

	expect( onChange ).not.toHaveBeenCalled();
} );
```

Reply:
Hi @aduth we should support negative numbers, the logic bug was solved and this test case and one to explicitly test negative ranges were added.